storage/database/src/key.rs

```diff
@@ -1,54 +1,175 @@
-        left.cmp(right)
-    }
+    const KEY_COMPARE: KeyCompare = KeyCompare::Lexicographic;
 }
 
 //---------------------------------------------------------------------------------------------------- Impl
-impl<T> Key for T
-where
-    T: Storable + Sized,
-{
-    type Primary = Self;
+/// [`Ord`] comparison for arrays/vectors.
+impl<const N: usize, T> Key for [T; N] where T: Key + Storable + Sized + bytemuck::Pod {}
+impl<T: bytemuck::Pod + Debug + Ord> Key for StorableVec<T> {}
+
+/// [`Ord`] comparison for misc types.
+///
+/// This is not a blanket implementation because
+/// it allows outer crates to define their own
+/// comparison functions for their `T: Storable` types.
+impl Key for () {}
+impl Key for StorableBytes {}
+impl Key for StorableStr {}
+
+/// Number comparison.
+///
+/// # Invariant
+/// This must _only_ be implemented for [`u32`], [`u64`] (and maybe [`usize`]).
+///
+/// This is because:
+/// 1. We use LMDB's `INTEGER_KEY`[^1] flag when this enum variant is used
+/// 2. LMDB only supports these types when using that flag[^2]
+///
+/// See: <https://docs.rs/heed/0.20.0-alpha.9/heed/struct.DatabaseFlags.html#associatedconstant.INTEGER_KEY>
+///
+/// Other numbers will still have the same behavior, but they use
+/// [`impl_custom_numbers_key`] and essentially pass LMDB a "custom"
+/// number compare function.
+macro_rules! impl_number_key {
+    ($($t:ident),* $(,)?) => {
+        $(
+            impl Key for $t {
+                const KEY_COMPARE: KeyCompare = KeyCompare::Number;
+            }
+        )*
+    };
+}
+
+impl_number_key!(u32, u64, usize);
+#[cfg(not(any(target_pointer_width = "32", target_pointer_width = "64")))]
+compile_error!("`cuprate_database`: `usize` must be equal to `u32` or `u64` for LMDB's `usize` key sorting to function correctly");
+
+/// Custom number comparison for other numbers.
+macro_rules! impl_custom_numbers_key {
+    ($($t:ident),* $(,)?) => {
+        $(
+            impl Key for $t {
+                // Just forward the the number comparison function.
+                const KEY_COMPARE: KeyCompare = KeyCompare::Custom(|left, right| {
+                    KeyCompare::Number.as_compare_fn::<$t>()(left, right)
+                });
+            }
+        )*
+    };
+}
+
+impl_custom_numbers_key!(u8, u16, u128, i8, i16, i32, i64, i128, isize);
+
+//---------------------------------------------------------------------------------------------------- KeyCompare
+/// Comparison behavior for [`Key`]s.
+///
+/// This determines how the database sorts [`Key`]s inside a database [`Table`](crate::Table).
+///
+/// See [`Key`] for more info.
+#[derive(Default, Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum KeyCompare {
+    /// [Lexicographical comparison](https://doc.rust-lang.org/1.79.0/std/cmp/trait.Ord.html#lexicographical-comparison),
+    /// i.e. a straight byte comparison.
+    ///
+    /// This is the default.
+    #[default]
+    Lexicographic,
```

Comment:
I think it might be better to call this `Default`, as lexicographic still puts a requirement on backends.

If this was instead `Default` we can just use any method for the backend, whichever is more performant.

Not really an issue now as no backend actually has another default than lexicographic.